File: examples/quest_example.cpp
/**
 * Maestro + QuEST Integration — C++ Example
 * ==========================================
 *
 * Demonstrates using the QuEST simulator backend through Maestro's
 * C interface. This is the C++ equivalent of quest_example.py.
 *
 * Covers:
 *   1. Initializing the QuEST library via Maestro's Factory
 *   2. Running a Bell-state circuit via QASM with SimpleExecute
 *   3. Estimating Pauli expectation values with SimpleEstimate
 *   4. GHZ State (4 qubits)
 *
 * Build:
 *   This example must be compiled against the Maestro library (libmaestro).
 *
 *   cmake -B build -DCMAKE_BUILD_TYPE=Release \
 *         -DMAESTRO_DIR=/path/to/maestro
 *   cmake --build build --target quest_example
 *
 * Run (set library paths so both libmaestro and libmaestroquest are found):
 *
 *   # macOS
 *   export DYLD_LIBRARY_PATH=/path/to/maestro/build:/path/to/maestro-quest-interface/build
 *   ./build/quest_example
 *
 *   # Linux
 *   export LD_LIBRARY_PATH=/path/to/maestro/build:/path/to/maestro-quest-interface/build
 *   ./build/quest_example
 */

// Only include Maestro's C interface — no C++ headers needed.
#include "maestrolib/Interface.h"

#include <cstdlib>
#include <iostream>

// Forward-declare the QuEST init function from libmaestro.
// This avoids pulling in the full Factory.h header chain
// (which requires Eigen, Boost, QCSim, Aer, etc.).
namespace Simulators {
class SimulatorsFactory {
 public:
  static bool InitQuestLibrary();
};
}  // namespace Simulators

// SimulatorType::kQuestSim   = 3
// SimulationType::kStatevector = 0
static constexpr int kQuestSim    = 3;
static constexpr int kStatevector = 0;

static const char *SEP =
    "------------------------------------------------------------";

// ----------------------------------------------------------------
// 1. Bell State via QASM (SimpleExecute)
// ----------------------------------------------------------------
void bell_state_qasm() {
    std::cout << SEP << "\n"
              << "1. Bell State via QASM (SimpleExecute)\n"
              << SEP << "\n";

    unsigned long int handle = CreateSimpleSimulator(2);
    RemoveAllOptimizationSimulatorsAndAdd(handle, kQuestSim, kStatevector);

    const char *qasm =
        "OPENQASM 2.0;\n"
        "include \"qelib1.inc\";\n"
        "qreg q[2];\n"
        "creg c[2];\n"
        "h q[0];\n"
        "cx q[0], q[1];\n"
        "measure q -> c;\n";

    const char *config = "{\"shots\": 1000}";

    char *result = SimpleExecute(handle, qasm, config);
    if (result) {
        std::cout << "   Result: " << result << "\n";
        FreeResult(result);
    } else {
        std::cerr << "   SimpleExecute returned null\n";
    }

    std::cout << "   ✓ Bell state executed\n\n";
    DestroySimpleSimulator(handle);
}

// ----------------------------------------------------------------
// 2. Expectation Values via QASM (SimpleEstimate)
// ----------------------------------------------------------------
void expectation_values_qasm() {
    std::cout << SEP << "\n"
              << "2. Expectation Values via QASM (SimpleEstimate)\n"
              << SEP << "\n";

    unsigned long int handle = CreateSimpleSimulator(2);
    RemoveAllOptimizationSimulatorsAndAdd(handle, kQuestSim, kStatevector);

    const char *qasm =
        "OPENQASM 2.0;\n"
        "include \"qelib1.inc\";\n"
        "qreg q[2];\n"
        "h q[0];\n"
        "cx q[0], q[1];\n";

    const char *observables = "ZZ;XX;YY";
    const char *config = "{}";

    char *result = SimpleEstimate(handle, qasm, observables, config);
    if (result) {
        std::cout << "   Result: " << result << "\n";
        std::cout << "   Expected: ZZ=+1, XX=+1, YY=-1\n";
        FreeResult(result);
    } else {
        std::cerr << "   SimpleEstimate returned null\n";
    }

    std::cout << "   ✓ Expectation values computed\n\n";
    DestroySimpleSimulator(handle);
}

// ----------------------------------------------------------------
// 3. GHZ State (4 qubits)
// ----------------------------------------------------------------
void ghz_state_qasm() {
    std::cout << SEP << "\n"
              << "3. GHZ State (4 qubits)\n"
              << SEP << "\n";

    unsigned long int handle = CreateSimpleSimulator(4);
    RemoveAllOptimizationSimulatorsAndAdd(handle, kQuestSim, kStatevector);

    const char *qasm =
        "OPENQASM 2.0;\n"
        "include \"qelib1.inc\";\n"
        "qreg q[4];\n"
        "creg c[4];\n"
        "h q[0];\n"
        "cx q[0], q[1];\n"
        "cx q[0], q[2];\n"
        "cx q[0], q[3];\n"
        "measure q -> c;\n";

    const char *config = "{\"shots\": 2000}";

    char *result = SimpleExecute(handle, qasm, config);
    if (result) {
        std::cout << "   Result: " << result << "\n";
        FreeResult(result);
    } else {
        std::cerr << "   SimpleExecute returned null\n";
    }

    std::cout << "   ✓ GHZ state executed\n\n";
    DestroySimpleSimulator(handle);
}

// ----------------------------------------------------------------
// Main
// ----------------------------------------------------------------
int main() {
    std::cout << "\n"
              << "============================================================\n"
              << "  Maestro + QuEST Integration — C++ Example\n"
              << "============================================================\n\n";

    // Initialize the Maestro engine
    void *maestro = GetMaestroObject();
    if (!maestro) {
        std::cerr << "Failed to initialize Maestro.\n";
        return 1;
    }

    // Initialize the QuEST library (loads libmaestroquest dynamically)
    bool questOk = Simulators::SimulatorsFactory::InitQuestLibrary();
    std::cout << "QuEST library initialized: " << (questOk ? "yes" : "no")
              << "\n\n";
    if (!questOk) {
        std::cerr << "Failed to load QuEST library. Make sure libmaestroquest "
                     "is on the library search path.\n";
        return 1;
    }

    bell_state_qasm();
    expectation_values_qasm();
    ghz_state_qasm();

    std::cout << SEP << "\n"
              << "All examples completed successfully!\n"
              << SEP << "\n";

    return 0;
}

File: examples/quest_example.py
"""
Maestro + QuEST Integration Example
====================================

Demonstrates using the QuEST quantum simulator backend through Maestro's
Python bindings.  Covers:

  1. Library initialisation and availability check
  2. Running a Bell-state circuit via QASM with simple_execute
  3. Estimating Pauli expectation values with simple_estimate
  4. Building a circuit programmatically with the QuantumCircuit API
"""

import maestro
from maestro.circuits import QuantumCircuit

SEPARATOR = "-" * 60


def check_quest_backend():
    """Initialize and verify the QuEST backend is available."""
    print(SEPARATOR)
    print("1. Initializing QuEST backend")
    print(SEPARATOR)

    ok = maestro.init_quest()
    print(f"   init_quest() returned: {ok}")
    print(f"   is_quest_available() : {maestro.is_quest_available()}")

    if not ok:
        raise RuntimeError(
            "QuEST library could not be loaded. "
            "Make sure libmaestroquest.so/.dylib is on DYLD_LIBRARY_PATH / LD_LIBRARY_PATH."
        )
    print("   ✓ QuEST backend ready\n")


def run_bell_state_qasm():
    """Execute a Bell-state circuit from a QASM string."""
    print(SEPARATOR)
    print("2. Bell State via QASM (simple_execute)")
    print(SEPARATOR)

    qasm = """\
OPENQASM 2.0;
include "qelib1.inc";
qreg q[2];
creg c[2];
h q[0];
cx q[0], q[1];
measure q -> c;
"""

    result = maestro.simple_execute(
        qasm,
        simulator_type=maestro.SimulatorType.QuestSim,
        simulation_type=maestro.SimulationType.Statevector,
        shots=1000,
    )

    print(f"   Counts       : {result['counts']}")
    print(f"   Time (s)     : {result['time_taken']:.6f}")
    print(f"   Simulator ID : {result['simulator']}")
    print(f"   Method ID    : {result['method']}")
    print("   ✓ Bell state executed successfully\n")
    return result


def estimate_expectations():
    """Estimate Pauli expectation values for a Bell state."""
    print(SEPARATOR)
    print("3. Expectation Values via QASM (simple_estimate)")
    print(SEPARATOR)

    qasm = """\
OPENQASM 2.0;
include "qelib1.inc";
qreg q[2];
h q[0];
cx q[0], q[1];
"""

    result = maestro.simple_estimate(
        qasm,
        "ZZ;XX;YY",
        simulator_type=maestro.SimulatorType.QuestSim,
        simulation_type=maestro.SimulationType.Statevector,
    )

    labels = ["ZZ", "XX", "YY"]
    print("   Observable  Expectation")
    for label, val in zip(labels, result["expectation_values"]):
        print(f"   {label:>10}  {val:+.6f}")
    print(f"   Time (s)  : {result['time_taken']:.6f}")
    print("   ✓ Expectation values computed\n")
    return result


def run_quantum_circuit_api():
    """Build and execute a GHZ state using the QuantumCircuit API."""
    print(SEPARATOR)
    print("4. GHZ State via QuantumCircuit API")
    print(SEPARATOR)

    n_qubits = 4
    qc = QuantumCircuit()
    qc.h(0)
    for i in range(1, n_qubits):
        qc.cx(0, i)
    qc.measure_all()

    result = qc.execute(
        simulator_type=maestro.SimulatorType.QuestSim,
        simulation_type=maestro.SimulationType.Statevector,
        shots=2000,
    )

    print(f"   {n_qubits}-qubit GHZ state")
    print(f"   Counts     : {result['counts']}")
    print(f"   Time (s)   : {result['time_taken']:.6f}")
    print("   ✓ GHZ circuit executed successfully\n")
    return result


def run_parametric_circuit():
    """Build a parameterized circuit with rotation gates."""
    print(SEPARATOR)
    print("5. Parametric Circuit (Rotation Gates)")
    print(SEPARATOR)

    import math

    qc = QuantumCircuit()
    # Prepare a variational ansatz-like circuit
    for q in range(3):
        qc.ry(q, math.pi / 4)
    qc.cx(0, 1)
    qc.cx(1, 2)
    for q in range(3):
        qc.rz(q, math.pi / 3)
    qc.measure_all()

    result = qc.execute(
        simulator_type=maestro.SimulatorType.QuestSim,
        simulation_type=maestro.SimulationType.Statevector,
        shots=2000,
    )

    print(f"   Counts     : {result['counts']}")
    print(f"   Time (s)   : {result['time_taken']:.6f}")
    print("   ✓ Parametric circuit executed successfully\n")
    return result


def main():
    print()
    print("=" * 60)
    print("  Maestro + QuEST Integration Example")
    print("=" * 60)
    print()

    check_quest_backend()
    run_bell_state_qasm()
    estimate_expectations()
    run_quantum_circuit_api()
    run_parametric_circuit()

    print(SEPARATOR)
    print("All examples completed successfully!")
    print(SEPARATOR)


if __name__ == "__main__":
    main()
